manager/drivers/secrets.go

```diff
@@ -0,0 +1,52 @@
+package drivers
+
+import (
+	"fmt"
+	"github.com/docker/docker/pkg/plugingetter"
+	"github.com/docker/swarmkit/api"
+)
+
+const (
+	// SecretsProviderAPI is the endpoint for fetching secrets from plugins
+	SecretsProviderAPI = "/SecretProvider.GetSecret"
+
+	// SecretsProviderCapability is the secrets provider plugin capability identification
+	SecretsProviderCapability = "secretprovider"
+)
+
+// SecretDriver provides secrets from different stores
+type SecretDriver struct {
+	plugin plugingetter.CompatPlugin
+}
+
+// NewSecretDriver creates a new driver that provides third party secrets
+func NewSecretDriver(plugin plugingetter.CompatPlugin) *SecretDriver {
+	return &SecretDriver{plugin: plugin}
+}
+
+// Get gets a secret from the secret provider
+func (d *SecretDriver) Get(spec *api.SecretSpec, serviceSpec api.ServiceSpec) ([]byte, error) {
+	secretReq := &SecretsProviderRequest{Name: spec.Annotations.Name, ServiceSpec: serviceSpec}
+	var secretResp SecretsProviderResponse
+	err := d.plugin.Client().Call(SecretsProviderAPI, secretReq, &secretResp)
+	if err != nil {
+		return nil, err
+	}
+	if secretResp.Err != "" {
+		return nil, fmt.Errorf(secretResp.Err)
+	}
+	// Assign the secret
+	return []byte(secretResp.Value), nil
+}
+
+// SecretsProviderRequest is the request specification for retrieving secrets from plugins.
+type SecretsProviderRequest struct {
+	Name        string          `json:"name"`        // Name is the name of the secret plugin
+	ServiceSpec api.ServiceSpec `json:"serviceSpec"` // ServiceSpec is the task's service specification
```

Comment:
@cpuguy83 I suppose the plugin is given the user specified name of the secret, which may resolve to different secret values for different services (e.g. different paths in the generic secret backend or different common names in a server cert in the context of a Vault-based plugin).